plugins/org.eclipse.cbi.p2repo.sbom/src/org/eclipse/cbi/p2repo/sbom/ClearlyDefinedApi.java

```diff
@@ -143,42 +137,30 @@ public CompletableFuture<Void> submitRequest(Component component, URI uri) {
 	 * @throws InterruptedException if the wait is interrupted
 	 */
 	public void waitForCompletion() throws InterruptedException {
-		// First wait for queue to drain and all active futures to complete
-		while (!requestQueue.isEmpty() || hasActiveFutures()) {
-			// Use a slightly longer sleep to reduce CPU usage in polling
+		// Wait for queue to drain and all active futures to complete
+		while (!requestQueue.isEmpty() || !activeFutures.isEmpty()) {
 			Thread.sleep(100);
 		}
 		
-		// Give a final moment for any in-flight requests to update state
-		if (!activeFutures.isEmpty()) {
-			CompletableFuture<?>[] futuresArray = activeFutures.keySet().toArray(new CompletableFuture<?>[0]);
-			if (futuresArray.length > 0) {
-				try {
-					CompletableFuture.allOf(futuresArray).get(30, TimeUnit.SECONDS);
-				} catch (Exception e) {
-					// Log but don't fail - some requests may have legitimately failed
-					System.err.println("Some ClearlyDefined requests did not complete: " + e.getMessage());
-				}
+		// Final wait for any remaining futures
+		CompletableFuture<?>[] remaining = activeFutures.keySet().toArray(new CompletableFuture<?>[0]);
+		if (remaining.length > 0) {
+			try {
+				CompletableFuture.allOf(remaining).get(30, TimeUnit.SECONDS);
+			} catch (Exception e) {
+				// Some requests may have legitimately failed
+				System.err.println("Some ClearlyDefined requests did not complete: " + e.getMessage());
```

Comment:
this is way to complicated, we can simply drain one element from the activefutures and join on it (without timeout) then go to the next.

also we need some kind of syncronization herw between the waiting and the background thread:

so it should be something like while(true) { syncronize on shared object) / check if requestquque is empty if yes, and active features is empty -> break, else -> loop thourgh active features and wait for each of them then restart (features will be removed from list automatically)

The queue thread needs to also syncronize to only remove item + add to active futures in one go so we do not have race condition where queue is empty but not already added to active futures!